src/goto-programs/goto_contractor.cpp

```diff
@@ -967,47 +989,42 @@ expr2tc interval_analysis_ibex_contractor::result_of_outer(expr2tc exp)
         continue;
 
       //if there is a lower bound,
-      if (!isinf(var.second.getInterval().lb()))
+      if (isnormal(var.second.getInterval().lb()))
       {
         //check if it overflows when cast back to integer
-        long integer;
-        if (
-          (long)var.second.getInterval().lb() > 0 &&
-          var.second.getInterval().lb() < 0)
-          integer = (long)var.second.getInterval().lb() + 1;
-        else
-          integer = (long)var.second.getInterval().lb();
+        BigInt r(0);
+        if (is_signedbv_type(X->type))
+        {
+          r = r.power2(X->type->get_width() - 1);
+          r = -r;
+        } // if its unsigned then its just zero
+
+        BigInt integerValue(0);
 
-        auto w = to_symbol2t(X).type->get_width();
-        int is_unsigned = is_unsignedbv_type(X->type) ? 0 : 1;
-        long long type_limit = -pow(2, w - is_unsigned);
-        if (integer < type_limit)
-          integer = type_limit;
+        if (var.second.getInterval().lb() < r.to_int64())
+          integerValue = r;
+        else
+          integerValue = (long)ceil(var.second.getInterval().lb());
 
-        auto lb = create_value_expr(integer, X->type);
+        auto lb = constant_int2tc(X->type, integerValue);
         auto cond1 = greaterthanequal2tc(X, lb);
         cond = and2tc(cond, cond1);
       }
-      if (!isinf(var.second.getInterval().ub()))
+      if (isnormal(var.second.getInterval().ub()))
       {
-        long integerValue;
-
-        if (
-          (long)var.second.getInterval().ub() < 0 &&
-          var.second.getInterval().ub() > 0)
-          integerValue = (long)var.second.getInterval().ub() - 1;
-        else
-          integerValue = (long)var.second.getInterval().ub();
+        BigInt r(0);
+        r = r.power2(
```

Comment:
`BigInt::power2` is a static function, no need to init to 0 first and call it as a member.